### ISSUE TITLE
0.142.0 - Error -> chrome_elf.dll missing

### ISSUE THREAD

**Issue by lvmajor** [microsoft/playwright-sharp#911]
Hey there, me again (sorry) :) 

I just updated the Nuget package to latest version (0.142.0) but now when I try to generate a pdf, I get a `PlaywrightSharpException: Host system is missing dependencies!` exception

I am using it like so: 
```
await Playwright.InstallAsync();
using var playwright = await Playwright.CreateAsync();
await using var browser = await playwright.Chromium.LaunchAsync(new LaunchOptions
{
    Headless = true
});
var page = await browser.NewPageAsync();
await page.SetContentAsync(htmlContent, LifecycleEvent.Load);
var bytesResponse = await page.GetPdfAsync(null, 1, true, "<style>body {font-size: 12px;}</style>Test header template", "Test footer template", true, false, "", null,
                null, null, new Margin { Bottom = "1in", Top = "2in" }, false);
```

**Comment by lvmajor** [comment]
Windows 10

**Comment by lvmajor** [comment]
It says `Note: use DEBUG=pw:api environment variable and rerun to capture Playwright logs.` is there an easy way for me to pass it to the actual command that launches the conversion so that I can see the Playwright logs, maybe that could help diagnose the issue here... ?

**Comment by kblok** [comment]
@lvmajor can you get me the windows build?
We don't have support for DEBUG yet.

**Comment by lvmajor** [comment]
![image](https://user-images.githubusercontent.com/1885400/95247198-8ed18b00-07e3-11eb-95c7-ab16b36e32e0.png)

**Comment by kblok** [comment]
Could you try going to this directory `%USERPROFILE%\AppData\Local` and remove the `ms-playwright` directory?

**Comment by lvmajor** [comment]
I just did, with no success sadly. Here is a screenshot of the `%USERPROFILE%\AppData\Local\ms-playwright\chromium-799411\chrome-win` after deleting and re-running the script above

![image](https://user-images.githubusercontent.com/1885400/95247542-fe477a80-07e3-11eb-83d9-81f9557d8a0f.png)

**Comment by glacasa** [comment]
I have the same problem and I came here to see if it was reported, I arrived a few minutes late ^^
I'm on the exact same Windows 10 version 

I'm trying to use PlaywrightSharp in an aspnet core project.
When I run it from Kestrel, it works perfectly, but when I use it in IIS Express or IIS I have the error.

**Comment by kblok** [comment]
@lvmajor is your scenario also IIS?

**Comment by lvmajor** [comment]
> I have the same problem and I came here to see if it was reported, I arrived a few minutes late ^^
> I'm on the exact same Windows 10 version
> 
> I'm trying to use PlaywrightSharp in an aspnet core project.
> When I run it from Kestrel, it works perfectly, but when I use it in IIS Express or IIS I have the error.

@kblok 
I'm using it with asp.net core (latest release in the 3.x release train), also on IIS Express. I'm really wondering though how running it in Kestrel makes it work....?

**Comment by kblok** [comment]
Kestrel runs under your user. IIS might user another user context.

**Comment by lvmajor** [comment]
But that would cause some permission issues, not missing dll no?

**Comment by kblok** [comment]
Correct. I'll try to replicate the issue.

**Comment by lvmajor** [comment]
^I just ran the same scenario launching the app via the console instead of via Visual Studio/IIS Express and it works.... so you were right it does look like it has something to do with IIS Express....

**Comment by lvmajor** [comment]
Hmmm... looks like it was luck in the first try ahah. Now it doesn't work either when running through console app
![image](https://user-images.githubusercontent.com/1885400/95249142-436cac00-07e6-11eb-91f2-f9abdd965e21.png)

**Comment by lvmajor** [comment]
Trying to investigate on my side, I did restart the app (via console) and now it works (multiple requests tested as I thought it might have been a file lock problem... but that doesn't seem to be the case now...)

After a minute or so though it stops to work...

**Comment by lvmajor** [comment]
Ok so I did test a couple times the fact of restarting the app and each and every time, it works for a minute or so before it stops working. 

So that doesn't seem to point to a permission problem as otherwise it would never work at all. It might be a problem caused by a file lock or something like that (maybe when the page/browser is closed, something is not gracefully shut down or something similar to that... I'm really shooting in the dark here, just trying to help figuring it out as I really need it ahah!)

**Comment by kblok** [comment]
@lvmajor can you share a piece of your code?
I’m interesting in how you load playwright, the browser and how you dispose them.

**Comment by lvmajor** [comment]
@kblok most of it is above in the original post. 

The only missing lines from the above snippet have been added and the whole thing looks like this:
```
private async Task<FileResult> ConvertHtmlToPdf(FileToRenderToPdf fileToRender = null, string fileDownloadName = "filename")
{
  await Playwright.InstallAsync();
  using var playwright = await Playwright.CreateAsync();
  await using var browser = await playwright.Chromium.LaunchAsync(new LaunchOptions
  {
    Headless = true,
    Args = new string[] {}
  });
  var page = await browser.NewPageAsync();
  var htmlContent = await File.ReadAllTextAsync(fileToRender.FilePath);
  await page.SetContentAsync(htmlContent, LifecycleEvent.Load);
  var bytesResponse = await page.GetPdfAsync(null, 1, true, "<style>body {font-size: 12px;}</style>Test header template", "Test footer template", true, false, "", null,
                  null, null, new Margin { Bottom = "1in", Top = "2in" }, false);

  FileResult fileResult = new FileContentResult(bytesResponse, "application/pdf");
  fileResult.FileDownloadName = fileDownloadName + ".pdf";
  return fileResult;
}
```

**Comment by lvmajor** [comment]
I am not actually closing the browser or page manually at all as I thought it was all handled automatically (also based on the examples, it does not seem required)

**Comment by kblok** [comment]
@lvmajor yeah, the `using` should do the trick.

**Comment by lvmajor** [comment]
@kblok I just saw you opened a PR to fix this, but I'm not 100% confident this is the only issue here as running the app in console also demonstrated the same behavior after a certain amount of time (around 1 minute according to the tests I did). Could that also be related to the dll not being loaded? In my understanding, it the dll really wasn't loaded, it would not even work the first time(s) I perform the same request no?

Here is the error log I get when running in console app
```log
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
PlaywrightSharp.PlaywrightSharpException: Host system is missing dependencies!

Full list of missing libraries:
    chrome_elf.dll

Note: use DEBUG=pw:api environment variable and rerun to capture Playwright logs.
   at PlaywrightSharp.Helpers.TaskHelper.WithTimeout[T](Task`1 task, TimeSpan timeout, Func`2 exceptionFactory) in /Users/neo/Documents/Coding/Personal/playwright-sharp/src/PlaywrightSharp/Helpers/TaskHelper.cs:line 139
   at PlaywrightSharp.Transport.Connection.SendMessageToServerAsync[T](String guid, String method, Object args, Boolean ignoreNullValues, JsonSerializerOptions options, Boolean treatErrorPropertyAsError) in /Users/neo/Documents/Coding/Personal/playwright-sharp/src/PlaywrightSharp/Transport/Connection.cs:line 220
   at PlaywrightSharp.BrowserType.LaunchAsync(LaunchOptions options) in /Users/neo/Documents/Coding/Personal/playwright-sharp/src/PlaywrightSharp/BrowserType.cs:line 102
   at PlaywrightSharp.Chromium.ChromiumBrowserType.LaunchAsync(LaunchOptions options) in /Users/neo/Documents/Coding/Personal/playwright-sharp/src/PlaywrightSharp/Chromium/ChromiumBrowserType.cs:line 18
   at <ProjectName>.Services.WkHtmlToPdfConverterService.ConvertHtmlToPdf(FileToRenderToPdf fileToRender, String fileDownloadName, PlaywrightSharpPdfOptions pdfOptions, String languageCode) in C:\Users\<LocalUser>\Documents\Repos\Apps\<ProjectName>\<ProjectName>\Services\WkHtmlToPdfConverterService.cs:line 242
   at <ProjectName>.Services.WkHtmlToPdfConverterService.ConvertHtmlToPdf(String filePath, Boolean renderWithRazorViewEngine, Object model, String fileDownloadName, PlaywrightSharpPdfOptions pdfOptions, String languageCode) in C:\Users\<LocalUser>\Documents\Repos\Apps\<ProjectName>\<ProjectName>\Services\WkHtmlToPdfConverterService.cs:line 217
   at <ProjectName>.Controllers.SampleReportsController.GenerateTrendingReportFile(String sampleNumber, Boolean renderHtml) in C:\Users\<LocalUser>\Documents\Repos\Apps\<ProjectName>\<ProjectName>\Controllers\SampleReportsController.cs:line 340
   at <ProjectName>.Controllers.SampleReportsController.GenerateReport(String sampleNumber, Boolean renderHtml, Nullable`1 showPreviousResults) in C:\Users\<LocalUser>\Documents\Repos\Apps\<ProjectName>\<ProjectName>\Controllers\SampleReportsController.cs:line 138
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at NToastNotify.NtoastNotifyAjaxToastsMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass5_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at SaasKit.Multitenancy.Internal.TenantResolutionMiddleware`1.Invoke(HttpContext context, ITenantResolver`1 tenantResolver)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.MigrationsEndPointMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 522.6836ms 500 text/html; charset=utf-8
```

**Comment by kblok** [comment]
Do you want to test the fix?
Unzip this file and copy the exe to `playwright-sharp-drivers\0.142.0.0` inside your bin folder.

[Uploading playwright-driver-win.exe.zip…]()

**Comment by lvmajor** [comment]
Yeah I will do that right away!

**Comment by lvmajor** [comment]
@kblok  Hmmm... Where is the actual file to download? And where should I put it exactly? In the bin folder of the ms-playwright install? (there's none of that I think...)

Thanks in advance!

**Comment by kblok** [comment]
The link is in my comment. This is the destination in my test project `PlaywrightSharpAspNetPlayground\bin\Debug\netcoreapp3.1\playwright-sharp-drivers\0.142.0.0` you should override an existing file.

**Comment by lvmajor** [comment]
The link posted has the following address: `https://github.com/microsoft/playwright-sharp/issues/911` (at least for me, maybe I don't see the same as you here.... not sure what's happening here but you can see in the following image what the link destination is when I hover over the link posted in your comment...)

![image](https://user-images.githubusercontent.com/1885400/95462798-fe5f8b80-0945-11eb-94b9-a9a53540c026.png)

**Comment by kblok** [comment]
Ops 😂 

[playwright-driver-win.exe.zip](https://github.com/microsoft/playwright-sharp/files/5348023/playwright-driver-win.exe.zip)

**Comment by lvmajor** [comment]
Just a quick comment while I'm testing, what is the reason behind the driver not being donwloaded in Release mode and only in Debug mode?

**Comment by kblok** [comment]
It's being generated in runtime. It's embedded inside the dll. It will download the appropriate driver for the platform it's running.

**Comment by lvmajor** [comment]
It seems to completely hang now... I will keep testing and report back as soon as possible

**Comment by kblok** [comment]
@lvmajor that's true. I know what's going on. I'll get you something :)

**Comment by kblok** [comment]
Could you try this?
[playwright-driver-win.exe.zip](https://github.com/microsoft/playwright-sharp/files/5348410/playwright-driver-win.exe.zip)

**Comment by lvmajor** [comment]
One thing is clear, there are some file handles that are not released properly when the app is closed as right now I can't replace the executable with the new one before killing the Node.js process that remains active after I killed the app (gracefully using ctrl+c in the console and waiting for it to shut down). But that is another thing and I don't think it's related here. 

I'll test the new one and report soon!

**Comment by kblok** [comment]
@lvmajor I experienced the same. These issues you don't know if it's you or Windows. but it worth checking that.

**Comment by lvmajor** [comment]
As of now it seems promising (currently running in console app only), will check with IIS soon

**Comment by lvmajor** [comment]
Yup that seems to work! I don't know what you changed exactly but I can run it under IIS now and it doesn't seem to crash. I also checked in the process explorer and the driver's process seems to be properly disposed after conversion is completed (as expected I guess... not sure if that's the case as I know there is a way to reuse same browser instance but I'm not completely sure how that is related to the driver...)

Do you expect releasing a patch version soon?

Thanks in advance!

**Comment by kblok** [comment]
@lvmajor as soon as my PR gets merged upstream I'll ship a v0.142.1 version with this fix.

**Comment by lvmajor** [comment]
All right, now hoping they won't wait a month to get it merged as it sometimes happen -_- 

Thanks for your help and have a great day! (Now I will go on their repo to put pressure on your PR to be merged ahah :) )
